[WIP/RFC] Make Parametric class constructor more like Julia

PhilipVinc opened the pull request:
I am putting this here to get some feedback...
This is a tentative implementation of what I described in #9

the objective was that given the parametric class

```python
@plum.parametric
class Test:
    def __init__(self, a, *, b=3):
        self.a = a
        self.b = b
```

I would like to replace 
```python
T = Test[list]
obj = T([1,2,3], b=4)
```
with
```python
obj = Test([1,2,3], b=4)
type(obj)  # __main__.Test[<class 'list'>]
```

The syntax `Test[par]` still works.

--

I achieve this by introducing a new MetaClass, that I call `ParametricTypeMeta`, which should be only used for parametric objects, which overrides `__getitem__` to call`__new__` instead of `__call__`, and I override `__call__` to only instantiate the object if it's a concrete subtype.

What would you think about this?

wesselb commented:
I think this looks great! The implementation looks sounds to me. I'd be very happy to like this merged!

Do you have any additional thoughts? Is a solution like this sufficient for your applications?

PhilipVinc commented:
This is sufficient for my needs for now.

It would be nice if the selection of which arguments are lifted to type parameters was overrideable (it wouldn't even be too hard. The `@parametric` decorator could check if a method like `__type_parameters(args) ->Tuple` was defined on the the class, and if so take it and use it to determine the type parameters, but... yeah. 
Let's leave this for the future.

Let me add some tests to this PR and bump the version and then I'd be super-happy if this can be merged (and a new release tagged)

wesselb commented:
> It would be nice if (...) Let's leave this for the future.

I had very similar thoughts, but agree that this is perhaps currently best left for the future.

> Let me add some tests to this PR and bump the version and then I'd be super-happy if this can be merged (and a new release tagged)

That sounds like a plan! I'll leave some comments on the PR. Moreover, should we adjust the README to explain this behaviour?

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
+        else:
+            T = cls
+
+        # this is the standard type.__call__ implementation
```
Shall we simplify this to `return type.__call__(T, *args, **kwargs)`?

PhilipVinc replied:
works a charm.

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
+
+        return self.__new__(self, *p)
+
+    def __call__(cls, *args, **kwargs):
```
The convention I've been using is to use `kw_args` rather than `kwargs`.

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
+    """Parametric Types must be instantiated with indexing."""
 
-class CovariantMeta(TypeMeta):
+    def __getitem__(self, *p):
```
Shall we name `p` to `ps` to indicate plurality?

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
+        # and then call the equivalent of T(arg1, arg2, **kwargs)
+
+        if not cls.is_concrete:
+            argsT = tuple(type(arg) for arg in args)
```
If `len(args) == 1`, is `argsT = args[0]` a sensible simplification? Moreover, shall we name `argsT` to `type_parameters` to slightly clarify?

PhilipVinc replied:
I agree with the renaming.
However I don't understand the semplification you are proposing. This code works for an arbitrary number of parameters

wesselb replied:
Currently, this is what happens:

```python
In [1]: from plum import parametric

In [2]: @parametric
   ...: class A:
   ...:     def __init__(self, *xs):
   ...:         pass
   ...:

In [3]: A(1)
Out[3]: <__main__.A[(<class 'int'>,)] at 0x7fbfa809bf60>

In [4]: A[int](1)
Out[4]: <__main__.A[<class 'int'>] at 0x7fbfa80b4a20>
```

Note that `A(1)` has `(int,)` rather than `int` as its type parameter. It's not necessarily an issue, but it may be nice to simplify `(int,)` to `int`.

PhilipVinc replied:
I thought it would work for the `()` constructor too. But evidently I was wrong..

I addressed the issue.

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
 
-class CovariantMeta(TypeMeta):
+    def __getitem__(self, *p):
+        # Type[Par1, Par2] creates a new parametric type
```
Would it make sense to move this comment and the below comment to the docstring of the class, so that the generated docs also include them?

wesselb commented on plum/parametric.py:
```diff
@@ -30,8 +30,37 @@
+        return obj
+
+    @property
+    def is_concrete(cls):
```
```suggestion
    def is_concrete(cls):
        """"bool: Check whether the parametric type is instantiated or not."""
```

wesselb reviewed (changes requested):
I've left some comments. Would you mind also running it through the `black` formatter?

PhilipVinc commented:
I addressed all review comments and added tests.

I had to go and change 2 tests in https://github.com/wesselb/plum/pull/11/commits/4200fcc5c34195be2eb2284a8fab6e0b33c45a74 because they where using the `()` syntax to instantiate the parametrised type, otherwise everything was smooth.

I also added a feature so that the following now throws an error
```
>>> Test[int]
<class '__main__.Test[<class 'int'>]'>
>>> Test[int][float]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/filippovicentini/Dropbox/Ricerca/Codes/Python/plum/plum/parametric.py", line 46, in __getitem__
    raise TypeError("Cannot specify type parameters. "
TypeError: Cannot specify type parameters. This type is concrete.
```

as opposed to current master where there is a weird behaviour
```python
>>> import plum
>>> @plum.parametric
... class Test:
...   pass
...
>>> Test[1]
<class '__main__.Test[1]'>
>>> Test[1][2]
<__main__.Test[1] object at 0x106f12d90>
>>> Test[1][2][3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Test[1]' object is not subscriptable
```

PhilipVinc commented:
I added a test for the branch raising an error so that coverage is back 100%

wesselb commented on plum/parametric.py:
```diff
@@ -31,7 +31,43 @@
+        # and then call the equivalent of T(arg1, arg2, **kw_args)
+
+        if not cls.is_concrete:
+            argsT = tuple(type(arg) for arg in args)
+            if len(argsT) == 1:
+                argsT = argsT[0]
+            T = cls[argsT]
```
```suggestion
            type_parameter = tuple(type(arg) for arg in args)
            if len(type_parameter) == 1:
                type_parameter = type_parameter[0]
            T = cls[type_parameter]
```

wesselb commented on plum/parametric.py:
```diff
@@ -31,7 +31,43 @@
+        else:
+            T = cls
+
+        # calls __new__ and __init__
```
```suggestion
        # Calls `__new__` and `__init__`.
```

wesselb commented on plum/parametric.py:
```diff
@@ -31,7 +31,43 @@
+            raise TypeError("Cannot specify type parameters. This type is concrete.")
+
+    def __call__(cls, *args, **kw_args):
+        # Type(arg1, arg2, kw_args) will first construct the
+        # parametric subtype T = Type[type(arg1), type(arg2)]
+        # and then call the equivalent of T(arg1, arg2, **kw_args)
```
```suggestion
        # `Type(arg1, arg2, **kw_args)` will first construct the
        # parametric subtype `T = Type[type(arg1), type(arg2)]`
        # and then call the equivalent of `T(arg1, arg2, **kw_args)`.
```

wesselb commented on plum/parametric.py:
```diff
@@ -31,7 +31,43 @@
 
-class CovariantMeta(TypeMeta):
+class ParametricTypeMeta(TypeMeta):
+    """Parametric Types can be instantiated with indexing.
```
```suggestion
    """Parametric types can be instantiated with indexing.
```

wesselb commented on tests/test_parametric.py:
```diff
@@ -75,16 +75,48 @@ def tuple_elements_are_identical(tup1, tup2):
     assert tuple_elements_are_identical(type_parameter(A[a1, a2]()), (a1, a2))
     assert tuple_elements_are_identical(type_parameter(A[1, a2]()), (1, a2))
 
+    # test raise error if specify parameters twice
```
```suggestion
    # Test that an error is raised if type parameters are specified twice.
```

wesselb reviewed (changes requested):
This looks great. Thanks for making the changes, adding a description to the README, and fixing up the coverage! I've added some minor style suggestions mostly  for consistency with the rest of the code base. Besides those, I'm happy to merge.

Perhaps one other feature which we could include is to simplify the printing of `a` in 

```python
In [1]: from plum import parametric

In [2]: @parametric
   ...: class A:
   ...:     def __init__(self, *args):
   ...:         pass
   ...:

In [3]:

In [3]: a = A(1, 2)

In [4]: a
Out[4]: <__main__.A[(<class 'int'>, <class 'int'>)] at 0x7f89b050c8d0>
```
to `<__main__.A[<class 'int'>, <class 'int'>] at 0x7f89b050c8d0>`, so without the surrounding `(` and `)`, which would amount to a small tweaking of [this line](https://github.com/wesselb/plum/blob/962b47f077c486e63467c841b29c2c3d281fb09b/plum/parametric.py#L82), but this can also be addressed later.

PhilipVinc commented:
I applied your suggestions and fixed the issue you raised about the name.
It was due to  `__getitem__` wrapping everything inside a tuple when it was already a tuple, before calling `__new__`.
It's fixed now (I pasted your code, which i thought i could get by without...)

wesselb commented on tests/test_parametric.py:
```diff
@@ -75,16 +75,48 @@ def tuple_elements_are_identical(tup1, tup2):
+
+    assert type_parameter(a1) == float
+    assert type_parameter(a2) == float
+    assert type(a1) == type(a2)
```
```suggestion
    assert type(a1) == type(a2)
    assert type(a1).__name__ == type(a2).__name__ == f"A[{float}]"
```

wesselb commented on tests/test_parametric.py:
```diff
@@ -75,16 +75,48 @@ def tuple_elements_are_identical(tup1, tup2):
-    assert a.x == 5.0
+    assert type_parameter(b1) == (float, int)
+    assert type_parameter(b2) == (float, int)
+    assert type(b1) == type(b2)
```
```suggestion
    assert type(b1) == type(b2)
    assert type(b1).__name__ == type(b2).__name__ == f"B[{float},{int}]"
```

wesselb reviewed (changes requested):
Ah, I see!

I've just suggested two more tests which test the printing, and then this looks good to go. :)

PhilipVinc commented:
good idea.
Sorry for being lazy.

wesselb commented:
No problem at all. This looks ready to merge then!

wesselb commented:
I've released the new version `v1.2.0` to PyPI. Thanks @PhilipVinc for the contribution!

PhilipVinc commented:
thanks a lot!
